packages/typespec-azure-resource-manager/src/rules/arm-resource-name-pattern.ts

```diff
@@ -0,0 +1,37 @@
+import { Program, createRule, getPattern } from "@typespec/compiler";
+
+import { getArmResources } from "../resource.js";
+
+/**
+ * Verify that a delete operation only
+ */
+export const armResourceNamePatternRule = createRule({
+  name: "arm-resource-name-pattern",
+  severity: "warning",
+  url: "https://azure.github.io/typespec-azure/docs/libraries/azure-resource-manager/rules/resource-name-pattern",
+  description: "The resource name parameter should be defined with a 'pattern' restriction.",
+  messages: {
+    default: `The resource name parameter should be defined with a 'pattern' restriction.`,
+  },
+  create(context) {
+    return {
+      root: (program: Program) => {
+        const resources = getArmResources(program);
+        for (const resource of resources) {
+          // find the name property
+          const nameProperty = [...resource.typespecType.properties.values()].find(
+            (p) => p.name === "name"
+          );
+          if (nameProperty !== undefined) {
+            const pattern = getPattern(program, nameProperty);
+            if (pattern === undefined) {
+              context.reportDiagnostic({
+                target: nameProperty,
+              });
+            }
```

Comment:
So, per the LintDiff rule it says it's also fine to apply the default ARM pattern here if one is not provided. We have a couple options:
1) don't do that and just always throw this warning
2) always apply the default if not specified, in which case this linter rule actually becomes a warning diagnostic and the logic would change
3) provide an easy mechanism to provide the default pattern and have the error message tell people to use that (like we did for the error codes rule pointing to other templates).

@markcowl @timotheeguerin @johanste any thoughts on what is preferrable?

Reply:
It would be preferable (to me) if there was an implicit default. But I'm not sure how many resources adhere to the same constraints.

Reply:
That rule could also be an easy code fix candidate so user can auto fix

Reply:
@timotheeguerin do we have documentation on how to do that?

Reply:
https://typespec.io/docs/extending-typespec/linters#provide-a-codefix

For linter in particular

Reply:
The auto suppress code fix in the compiler is a simple one that is similar in the logic

Reply:
Also feels like adding a decorator on a node feels like a common code fix we should maybe build a common helper piece in the same way we have the update identfier codefix

Reply:
Agreed, a codefix with the default value is the best way to document this.  In general, I think we should make a codefix an acceptance criterion for new linting rules, where that is possible.

Reply:
But, please, file an issue for adding the codefix for this, it is important we get this rule in.